src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -352,60 +404,501 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+
+	  / Store block in directory /
+	  rgw::d4n::BlockDirectory* tempBlockDir = source->driver->get_block_dir(); // remove later -Sam
+
+	  source->driver->get_cache_block()->hostsList.push_back(tempBlockDir->get_addr().host + ":" + std::to_string(tempBlockDir->get_addr().port)); // local cache address -Sam 
+	  source->driver->get_cache_block()->size = source->get_obj_size();
+	  source->driver->get_cache_block()->cacheObj.bucketName = source->get_bucket()->get_name();
+	  source->driver->get_cache_block()->cacheObj.objName = source->get_key().get_oid();
+
+	  int setDirReturn = tempBlockDir->set_value(source->driver->get_cache_block());
+
+	  if (setDirReturn < 0) {
+	    ldpp_dout(dpp, 20) << "D4N Filter: Block directory set operation failed." << dendl;
+	  } else {
+	    ldpp_dout(dpp, 20) << "D4N Filter: Block directory set operation succeeded." << dendl;
+	  }
+	}
+      }
+    }
+  }
+
+  if (ret < 0) 
+    ldpp_dout(dpp, 20) << "D4N Filter: Cache iterate operation failed." << dendl;
+
+  return next->iterate(dpp, ofs, end, cb, y); */
+}
+
+int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::flush_last_part()
+{
+  last_part = true;
+  return handle_data(bl_rem, 0, bl_rem.length());
+}
+
+int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl, off_t bl_ofs, off_t bl_len)
+{
+  auto rgw_get_obj_max_req_size = g_conf()->rgw_get_obj_max_req_size;
+
+  if (!last_part && bl.length() <= rgw_get_obj_max_req_size) {
+    auto r = client_cb->handle_data(bl, bl_ofs, bl_len);
+
+    if (r < 0) {
+      return r;
+    }
+  }
+
+  //Accumulating data from backend store into rgw_get_obj_max_req_size sized chunks and then writing to cache
+  if (write_to_cache) {
+    const std::lock_guard l(d4n_get_data_lock);
+    rgw::d4n::CacheBlock block;
+    rgw::d4n::BlockDirectory* blockDir = source->driver->get_block_dir();
+    block.hostsList.push_back(blockDir->get_addr().host + ":" + std::to_string(blockDir->get_addr().port));
+    block.cacheObj.bucketName = source->get_bucket()->get_name();
+    block.cacheObj.objName = source->get_key().get_oid();
+
+    if (bl.length() > 0 && last_part) { // if bl = bl_rem has data and this is the last part, write it to cache
+      std::string oid = this->oid + "_" + std::to_string(ofs) + "_" + std::to_string(bl_len);
+      block.size = bl.length();
+      block.blockId = ofs;
+      uint64_t freeSpace = filter->get_cache_driver()->get_free_space(dpp);
+      while(freeSpace < block.size) {
+        freeSpace += filter->get_policy_driver()->cachePolicy->eviction(dpp, filter->get_cache_driver());
+      }
+      if (filter->get_cache_driver()->put_async(dpp, oid, bl, bl.length(), source->get_attrs()) == 0) {
+        filter->get_policy_driver()->cachePolicy->update(dpp, oid, ofs, bl.length(), filter->get_cache_driver());
+        /* Store block in directory */
```

Comment:
If the block exists, shouldn't we be updating it? For example, we should append the local cache address to its host list.

Reply:
Yes, if the block exists in the directory, we will update the hosts list with the local host (again this applies when remote caches are available), you have done that in your PR and thats fine.

Reply:
Additionally, we shouldn't be storing any attrs with blocks here. The idea is to fetch the attrs from the backend store for the object, instead of storing all the attrs with each block. We don't cache the head object (which contains the attrs). This needs to be fixed.

Reply:
Additionally, if the block exists, should the rest of the block's attributes be updated in addition to the host list? For example, the block ID and length.